src/Storages/MergeTree/ReplicatedMergeTreeAttachThread.cpp

```diff
@@ -0,0 +1,213 @@
+#include <Storages/MergeTree/ReplicatedMergeTreeAttachThread.h>
+#include <Storages/StorageReplicatedMergeTree.h>
+
+namespace DB
+{
+
+namespace ErrorCodes
+{
+    extern const int ABORTED;
+}
+
+ReplicatedMergeTreeAttachThread::ReplicatedMergeTreeAttachThread(StorageReplicatedMergeTree & storage_)
+    : storage(storage_)
+    , log_name(storage.getStorageID().getFullTableName() + " (ReplicatedMergeTreeAttachThread)")
+    , log(&Poco::Logger::get(log_name))
+{
+    task = storage.getContext()->getSchedulePool().createTask(log_name, [this] { run(); });
+    const auto storage_settings = storage.getSettings();
+    retry_period = storage_settings->initialization_retry_period.totalSeconds();
+}
+
+ReplicatedMergeTreeAttachThread::~ReplicatedMergeTreeAttachThread()
+{
+    shutdown();
+}
+
+void ReplicatedMergeTreeAttachThread::shutdown()
+{
+    if (!shutdown_called.exchange(true))
+        task->deactivate();
+}
+
+void ReplicatedMergeTreeAttachThread::run()
+{
+    try
+    {
+        LOG_INFO(log, "Table will be in readonly mode until initialization is finished");
+        zookeeper = storage.tryGetZooKeeper();
+        if (!zookeeper)
+        {
+            // if we failed to connect to ZK cluster once we don't want to wait connection timeout twice
+            notifyIfFirstTry();
+            tryReconnect();
+        }
+
+        if (shutdown_called)
+        {
+            LOG_WARNING(log, "Shutdown called, cancelling initialization");
+            return;
+        }
+
+        const auto & zookeeper_path = storage.zookeeper_path;
+        bool metadata_exists = withRetries([&] { return zookeeper->exists(zookeeper_path + "/metadata"); });
+        if (!metadata_exists)
+        {
+            LOG_WARNING(log, "No metadata in ZooKeeper for {}: table will stay in readonly mode.", zookeeper_path);
+            storage.has_metadata_in_zookeeper = false;
+            finalizeInitialization();
+
+            notifyIfFirstTry();
+            return;
+        }
+
+        storage.has_metadata_in_zookeeper = true;
+
+        auto metadata_snapshot = storage.getInMemoryMetadataPtr();
+
+        const auto & replica_path = storage.replica_path;
+        /// May it be ZK lost not the whole root, so the upper check passed, but only the /replicas/replica
+        /// folder.
+        bool replica_path_exists = withRetries([&] { return zookeeper->exists(replica_path); });
+        if (!replica_path_exists)
+        {
+            LOG_WARNING(log, "No metadata in ZooKeeper for {}: table will stay in readonly mode", replica_path);
+            storage.has_metadata_in_zookeeper = false;
+            finalizeInitialization();
+
+            notifyIfFirstTry();
+            return;
+        }
+
+        /// In old tables this node may missing or be empty
+        String replica_metadata;
+        const bool replica_metadata_exists = withRetries([&] { return zookeeper->tryGet(replica_path + "/metadata", replica_metadata); });
```

Comment:
Such tables are probably too old and we can remove compatibility with them and throw an exception that says that upgrade from 20.4 and older to 22.9 and newer should be done through an intermediate version.

Reply:
maybe we can remove this in a different PR to focus on the main part of this PR?

Reply:
Yes, we can do it in separate PR. Btw, removal of this code will fix a bug: https://github.com/ClickHouse/ClickHouse/pull/30826#issuecomment-1220610613